### PR TITLE
[FLINK-14363][runtime] Prevent vertex from being affected by outdated deployment

### PR DESCRIPTION
**Pull request by zhuzhurk**
## What is the purpose of the change

DefaultScheduler currently will cancel latest execution of a vertex when the vertex version is outdated. This may lead to undesirable failover of a healthy vertex.

This PR is to remove the vertex cancel process in DefaultScheduler#stopDeployment. This should be valid because the cancellation is not not needed here, since the version of a scheduled vertex is only outdated after the vertex is canceled by others.

## Brief change log

  - *Removed the vertex cancellation in DefaultScheduler#stopDeployment*
  - *hotfix changes see each commit*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added unit test vertexIsNotAffectedByOutdatedDeployment in DefaultSchedulerTest*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/TestingLogicalSlot.java [hunk 1]
```diff
@@ -111,10 +113,14 @@ public Payload getPayload() {
 
 	@Override
 	public CompletableFuture<?> releaseSlot(@Nullable Throwable cause) {
-		slotOwner.returnLogicalSlot(this);
+		if (!isReleased) {
```
Comment (GJL) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/TestingLogicalSlot.java]: Is this change needed for your new test?

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java [hunk 2]
```diff
@@ -335,15 +336,51 @@ public void scheduleOnlyIfVertexIsCreated() throws Exception {
...
+		final SchedulingExecutionVertex sv1 = scheduler.getSchedulingTopology().getVertices().iterator().next();
+
+		// fail v1 and let it recover to SCHEDULED
+		scheduler.updateTaskExecutionState(new TaskExecutionState(
+			jobGraph.getJobID(),
+			v1.getCurrentExecutionAttempt().getAttemptId(),
+			ExecutionState.FAILED));
+		taskRestartExecutor.triggerScheduledTasks();
+
+		// fail v2 to finish all pending slot requests of the last round
```
Comment (GJL) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java]: I don't understand this comment or the test. `v1` and `v2` are completely independent from each other since there is only one job vertex in this job.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java [hunk 3]
```diff
@@ -412,13 +412,17 @@ private static Throwable maybeWrapWithNoResourceAvailableException(final Throwab
...
-		cancelExecutionVertex(deploymentHandle.getExecutionVertexId());
+	private void releaseUnassignedSlotIfPresent(final DeploymentHandle deploymentHandle) {
 		// Canceling the vertex normally releases the slot. However, we might not have assigned
 		// the slot to the vertex yet.
+		// Only release unassigned slot to guarantee no vertex state change happens here.
 		deploymentHandle
 			.getLogicalSlot()
-			.ifPresent(logicalSlot -> logicalSlot.releaseSlot(null));
+			.ifPresent(logicalSlot -> {
+				if (logicalSlot.getPayload() != null) {
```
Comment (GJL) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java]: Would it actually be wrong to omit this check?

Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java]: I gave some more thoughts on it and thinks we can even remove this release logic since an unassigned slot will never get released here.
- if a slot is assigned with a payload (iff it is assigned to an execution or is released), there's no need to release it here
- if a slot is not assigned and needs a release 
 a) in `assignResourceOrHandleError`'s outdated deployment handling block. It's not possible to happen because the vertex is outdated iff it has been restarted and the slot request will be canceled then
 b) in `deployOrHandleError`'s outdated deployment handling block. This happens iff the preceding `assignResourceOrHandleError` of the same vertex is done without a successful assigning, which means 
   b.1) case *a)* happened but it's not possible to happen
   b.2) unexpected error happened in `assignResourceOrHandleError`. But `deployOrHandleError` can not be invoked in this case since `deployAll` will propagate the unexpected error to force a JM restart (ignoring `deployIndividually` which would be removed in FLINK-14162, see discussion [here](https://github.com/apache/flink/pull/9860#discussion_r334367314)

Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java]: In simple, a slot that completes slot request normally aways is always assigned to an execution in `DefaultScheduler#assignResourceOrHandleError`, unless there is an unexpected error(including assigning failure) which will shut down the JM JVM.

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/TestingLogicalSlot.java (hunk 1 above)
Comment (GJL) [reply flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/TestingLogicalSlot.java]: ok

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java (hunk 2 above)
Comment (GJL) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java]: Ok, understood.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java (hunk 3 above)
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java]: Yes. For `deployIndividually()` the release logic for outdated deployment is also not needed. (`deployIndividually()` has a problem to not handle the unexpected error correctly, this is a different issue that can be fixed in FLINK-14162).

> However, the test can be kept.

If you mean the `skipDeploymentIfVertexVersionOutdated` test case, agreed we still need it.

Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java]: done.

Comment (GJL) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java]: I meant the new test `vertexIsNotAffectedByOutdatedDeployment`, which you also have kept.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java [hunk 4]
```diff
@@ -355,15 +356,51 @@ public void handleGlobalFailure() {
...
+		testExecutionSlotAllocator.disableAutoCompletePendingRequests();
+		final DefaultScheduler scheduler = createSchedulerAndStartScheduling(jobGraph);
+
+		final Iterator<ArchivedExecutionVertex> vertexIterator = scheduler.requestJob().getAllExecutionVertices().iterator();
+		final ArchivedExecutionVertex v1 = vertexIterator.next();
+		final ArchivedExecutionVertex v2 = vertexIterator.next();
+
+		final SchedulingExecutionVertex sv1 = scheduler.getSchedulingTopology().getVertices().iterator().next();
+
+		// fail v1 and let it recover to SCHEDULED
```
Comment (GJL) [inline flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java]: maybe add as a comment that the initial deployment of v1 will be outdated

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/TestingLogicalSlot.java [hunk 5]
```diff
@@ -56,6 +56,8 @@
 	@Nullable
 	private final SlotSharingGroupId slotSharingGroupId;
 
+	private boolean released = false;
```
Comment (GJL) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/TestingLogicalSlot.java]: The default value is already false.

File: flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java (hunk 4 above)
Comment (zhuzhurk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java]: Ok. Done.

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/TestingLogicalSlot.java (hunk 5 above)
Comment (zhuzhurk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/TestingLogicalSlot.java]: done.

**Comment by GJL** [comment]
Merging.
